core/src/main/scala/zio/schema/codec/ProtobufCodec.scala

```diff
@@ -0,0 +1,485 @@
+import java.nio.{ ByteBuffer, ByteOrder }
+
+import zio.stream.ZTransducer
+import zio.schema._
+import zio.schema.codec.Codec
+import zio.{ Chunk, ZIO, ZManaged }
+
+// TODO safe splitting of chunks?
+object ProtobufCodec extends Codec {
+  override def encoder[A](schema: Schema[A]): ZTransducer[Any, Nothing, A, Byte] =
+    ZTransducer(ZManaged.succeed(chunk => ZIO.succeed(Encoder.encodeChunk(schema, chunk))))
+
+  override def decoder[A](schema: Schema[A]): ZTransducer[Any, String, Byte, A] =
+    ZTransducer(ZManaged.succeed(chunk => ZIO.fromEither(Decoder.decodeChunk(schema, chunk))))
+
+  object Protobuf {
+
+    sealed trait WireType {}
+
+    case object VarInt                     extends WireType
+    case object Bit64                      extends WireType
+    case class LengthDelimited(width: Int) extends WireType
+    case object StartGroup                 extends WireType
+    case object EndGroup                   extends WireType
+    case object Bit32                      extends WireType
+
+    def flatFields(
+      structure: Map[String, Schema[_]],
+      nextFieldNumber: Int = 1
+    ): Map[Int, (String, Schema[_])] =
+      structure.toSeq
+        .foldLeft((nextFieldNumber, Map[Int, (String, Schema[_])]())) { (numAndMap, fieldAndSchema) =>
+          nestedFields(fieldAndSchema._1, fieldAndSchema._2, nextFieldNumber) match {
+            case Some(fields) => (numAndMap._1 + fields.size, numAndMap._2 ++ fields)
+            case None         => (numAndMap._1 + 1, numAndMap._2 + (numAndMap._1 -> fieldAndSchema))
+          }
+        }
+        ._2
+
+    private def nestedFields(
+      baseField: String,
+      schema: Schema[_],
+      nextFieldNumber: Int
+    ): Option[Map[Int, (String, Schema[_])]] = schema match {
+      case _: Schema.Record      => None
+      case _: Schema.Sequence[_] => None
+      case Schema.Enumeration(structure) =>
+        Some(flatFields(structure, nextFieldNumber)).map(
+          _.map {
+            case (fieldNumber, fieldAndSchema) =>
+              val field = fieldAndSchema._1
+              (fieldNumber,
+               (baseField,
+                Schema.Transform(fieldAndSchema._2.asInstanceOf[Schema[Any]],
+                                 (a: Any) => Right(Map(field -> a)),
+                                 (b: Map[String, Any]) => b.get(field).toRight("Missing value"))))
+          }
+        )
+      case Schema.Transform(codec, f, g) =>
+        nestedFields(baseField, codec, nextFieldNumber).map(_.map {
+          case (fieldNumber, fieldAndSchema) =>
+            (fieldNumber, (fieldAndSchema._1, Schema.Transform(fieldAndSchema._2.asInstanceOf[Schema[Any]], f, g)))
+        })
+      case _: Schema.Primitive[_] => None
+      case _: Schema.Tuple[_, _]  => None
+      case _: Schema.Optional[_]  => None
+    }
+  }
+
+  object Encoder {
+    import Protobuf._
+
+    def encodeChunk[A](schema: Schema[A], chunk: Option[Chunk[A]]): Chunk[Byte] =
```

Comment:
What is the `Option[Chunk[A]]`?

Reply:
Well, I clearly didn't know how to create a `ZTransducer` properly. So I used a (unneeded) `ZManaged` and the `apply` method.  But having another look now it's the types that kinda force me. I cannot use the `ZTransducer.fromFunction[M]` here as the `Codec` trait tells me to return a `ZTransducer[Any, Nothing, A, Byte]` (encoder) and `ZTransducer[Any, String, Byte, A]` (decoder). Now would it make sense to use `Chunk[Byte]`s here as well? It's not like I can transform a single `Byte` into a value `A`.